### DISCUSSION TITLE
Question about cancellation during buildDocument

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1683] [Q&A] [answered]
Hello,

When I edit a big document (300 000 lines) with my dsl it takes around 3.3s to build (from parsing to validation). I noticed that the cancellation of the previous build happen very late in case of multiple editions.

I instrumented the builder this way:
```ts
  protected override async buildDocuments(documents: LangiumDocument[], options: BuildOptions, cancelToken: Cancellation.CancellationToken): Promise<void> {
        try {
            console.time('buildDocuments');
            await super.buildDocuments(documents, options, cancelToken);
        }
        catch (e) {
            console.log('exception');
            throw e;
        }
        finally {
            console.timeEnd('buildDocuments');
        }
    }
```

If I append 10 characters in my file, one each 0.3 second (approximately), I have the following log:

```
exception
buildDocuments: 2.634s
exception
buildDocuments: 2.446s
exception
buildDocuments: 2.398s
exception
buildDocuments: 2.431s
exception
buildDocuments: 2.729s
exception
buildDocuments: 2.329s
exception
buildDocuments: 2.369s
exception
buildDocuments: 2.408s
exception
buildDocuments: 2.727s
buildDocuments: 3.351s
```
The exception raised by the CancellationToken happen after around 2.5s.

I tried to reduce the interruptionPeriod with `setInterruptionPeriod` but it didn't changed anything.

In such case the final document is built after more than 20s after the last edition.
Is there a way to interrupt the build quickly ?

**Comment by msujew** [comment, marked as answer]
Hey @ydaveluy,

Unfortunately, the first step in the document building pipeline is always parsing which is usually synchronous and therefore cannot be interrupted.

We built the `AsyncParser` for this, but it's not ideal. It has a fairly slow startup time, but you might want to use it anyway. See [here](https://github.com/eclipse-langium/langium/blob/main/packages/langium/test/parser/worker-thread-async-parser.test.ts) on how to use it.
